### DISCUSSION TITLE
[Auto Import] Better auto-import priority (Lower the priority of superseded modules)

### DISCUSSION THREAD

**Discussion by BlueGlassBlock** [microsoft/pylance-release#3328] [Enhancement]
According to this [doc site](https://docs.python.org/3.10/library/superseded.html), lots of modules are superseded / deprecated.

However, lots of them **still** have high priority and break users' auto-import experience.

For example:
type `from typing imp` and the first entry is `imp` module.
`Option`'s auto-import entry is from `optparse` rather than my local module.

And some related annoying imports:
`from ctypes import Union` rather than `from typing import Union` (also `typing.cast`)
`from re import I` (when completing `from ... import ...` too quickly)

I believe the auto-import problem has been lingering for quite a long time.

Is it possible to address a better algorithm (context-aware / prediction based on user data) to improve the state?

If not, could you please add a way for us to filter out those modules we don't want? (For example `this`, from #2535)

**Comment by ProgramRipper** [comment]
That's why I choose PyCharm...
Just kidding, sincerely hope that your great job could enhance it.

**Comment by luabud** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by ccrvlh** [comment]
I experience something similar, that I think is also related to auto suggestion logic. Normally when typing `cast`  the `ctypes` suggestion is usually before the `typing` suggestion. This is common with different packages as well. 

Other thing that I found, was that suggestions won't prioritize local packages, I have a package called `BaseService` that is used a lot across the application (a lot of modules import it), and every time I start typing `BaseService` always (100% of the times) recommends `BaseServer` from `socketserver` which is a package that I've never used before and doesn't appear in my project anywhere.

Maybe it could make sense to have a higher ranking for local objects (local in the sense of the user's project instead of third party libs)? Other option is maybe to keep track of import frequency (objects that were imported a lot have a lower threshold to appear in suggestions)? Besides the deprecated being lower ranked, maybe most commonly used imports (from Github's Python's codebases perhaps) could also be higher ranked? Examples: `packaging` appears before `pandas` and `numbers` appears before `numpy`, should it (honest question)? I don't know which is more popular.

At last, I also found that quick actions suggestions (`Cmd` + `.`) don't really work very well when doing something like `from pandas import ....` it won't show the available objects. A lot of libs have `__all__` which maybe can help?

Anyways, just ideas.

**Comment by rayansostenes** [comment]
Another auto-import problem that I always face is when I need to import something like `Self` it suggests the import from `typing` even tough I'm using python 3.9 where this is not yet available, I never got it to suggest the import from `typing_extensions`

![image](https://github.com/microsoft/pylance-release/assets/885388/53460922-a291-4ef2-a568-3590f8f0103b)

![image](https://github.com/microsoft/pylance-release/assets/885388/45f33aeb-9f05-45c6-b05d-364f5ed1aa72)

**Reply by DanielRosenwasser** [reply]
I actually filed an issue about suggesting newer features from `typing` over `typing_extensions`: https://github.com/microsoft/pylance-release/issues/4993

**Reply by heejaechang** [reply]
we have open PR to address this. @bschnurr do you have ETA for your PR?
